Allow images to be created on demand via the imagestreamtag api

smarterclayton opened the pull request:
A new subresource `imagestreamtags/copy` is added which accepts an ImageStreamTagCopy. This describes a request to create a new image from the provided docker image metadata (command, labels, env, ports, etc) and provide a single layer via direct submission, dramatically simplifying the act of creating an image. The caller can create a new from scratch image or add a new layer on top of an existing image. The endpoint then talks to the registry and abstracts the necessary details like creating a manifest, uploading the blob contents, and returning the new image.

The API is intended for use with binary uploads and other ad-hoc image composition.

smarterclayton commented:
Still WIP, tentatively targeted for 3.7

smarterclayton commented:
```
I0611 16:45:42.342377   10669 wrap.go:42] POST /apis/image.openshift.io/v1/namespaces/default/imagestreamtags/test:bar/clone: (3.298344ms) 201 [[oc/v1.6.1+5115d708d7 (darwin/amd64) kubernetes/010d313] 192.168.1.105:51585]
I0611 16:45:42.342805   10669 factory.go:185] Image stream "test" updated.
I0611 16:45:42.342840   10669 imagestream_controller.go:145] Queued import of stream default/test...
I0611 16:45:42.342848   10669 image_trigger_controller.go:383] Started syncing image stream "default/test"
I0611 16:45:42.342861   10669 image_trigger_controller.go:385] Finished syncing image stream "default/test" (15.977µs)
I0611 16:45:42.347142   10669 wrap.go:42] GET /api/v1/namespaces/default: (859.859µs) 200 [[openshift/v1.6.1+5115d708d7 (darwin/amd64) kubernetes/010d313] 192.168.1.105:63429]
I0611 16:45:42.347786   10669 wrap.go:42] GET /oapi/v1/users/~: (1.984949ms) 200 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.348242   10669 wrap.go:42] GET /api/v1/namespaces/default/services/kubernetes: (696.606µs) 200 [[openshift/v1.6.1+5115d708d7 (darwin/amd64) kubernetes/010d313] 192.168.1.105:63429]
I0611 16:45:42.358024   10669 wrap.go:42] GET /oapi/v1/users/~: (8.727905ms) 200 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.360322   10669 wrap.go:42] POST /oapi/v1/namespaces/default/localsubjectaccessreviews: (1.716416ms) 201 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.362271   10669 wrap.go:42] POST /oapi/v1/namespaces/default/localsubjectaccessreviews: (1.425999ms) 201 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.366326   10669 wrap.go:42] GET /oapi/v1/users/~: (1.233398ms) 200 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.368374   10669 wrap.go:42] POST /oapi/v1/namespaces/default/localsubjectaccessreviews: (1.577408ms) 201 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.370344   10669 wrap.go:42] POST /oapi/v1/namespaces/default/localsubjectaccessreviews: (1.478365ms) 201 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.377686   10669 clone.go:496] Wrote blob of 192 bytes with digests layer=sha256:051652a014c06bb792007f2eefc56909880c60d26813929d261d93fdce5a5fa1 blob=sha256:051652a014c06bb792007f2eefc56909880c60d26813929d261d93fdce5a5fa1
I0611 16:45:42.379753   10669 wrap.go:42] GET /oapi/v1/users/~: (1.118196ms) 200 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.381727   10669 wrap.go:42] POST /oapi/v1/namespaces/default/localsubjectaccessreviews: (1.211514ms) 201 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.383204   10669 wrap.go:42] POST /oapi/v1/namespaces/default/localsubjectaccessreviews: (1.128543ms) 201 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.387816   10669 wrap.go:42] GET /oapi/v1/users/~: (1.095776ms) 200 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.391142   10669 wrap.go:42] POST /oapi/v1/namespaces/default/localsubjectaccessreviews: (2.673396ms) 201 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.396671   10669 wrap.go:42] GET /oapi/v1/users/~: (1.094566ms) 200 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.398570   10669 wrap.go:42] POST /oapi/v1/namespaces/default/localsubjectaccessreviews: (1.375844ms) 201 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.401629   10669 wrap.go:42] GET /oapi/v1/users/~: (1.037526ms) 200 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.403139   10669 wrap.go:42] POST /oapi/v1/namespaces/default/localsubjectaccessreviews: (1.162792ms) 201 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.404562   10669 wrap.go:42] POST /oapi/v1/namespaces/default/localsubjectaccessreviews: (1.101572ms) 201 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.408460   10669 wrap.go:42] GET /oapi/v1/users/~: (1.04311ms) 200 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.409986   10669 wrap.go:42] POST /oapi/v1/namespaces/default/localsubjectaccessreviews: (1.148814ms) 201 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.411404   10669 wrap.go:42] POST /oapi/v1/namespaces/default/localsubjectaccessreviews: (1.091018ms) 201 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.415020   10669 wrap.go:42] GET /oapi/v1/users/~: (1.057558ms) 200 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.416523   10669 wrap.go:42] POST /oapi/v1/namespaces/default/localsubjectaccessreviews: (1.15025ms) 201 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.417975   10669 wrap.go:42] POST /oapi/v1/namespaces/default/localsubjectaccessreviews: (1.118963ms) 201 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.422262   10669 wrap.go:42] GET /oapi/v1/users/~: (1.000059ms) 200 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.423759   10669 wrap.go:42] POST /oapi/v1/namespaces/default/localsubjectaccessreviews: (1.14641ms) 201 [[dockerregistry/v3.6.0 (darwin/amd64) openshift/4321cf4] 192.168.1.105:63442]
I0611 16:45:42.429215   10669 clone.go:694] Persisted image stream: &api.ImageStream{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"test", GenerateName:"", Namespace:"default", SelfLink:"", UID:"4e48e8c9-4ed5-11e7-a403-7831c1b76042", ResourceVersion:"1986", Generation:1, CreationTimestamp:v1.Time{Time:time.Time{sec:63632803170, nsec:649107362, loc:(*time.Location)(0xfe9db60)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:""}, Spec:api.ImageStreamSpec{LookupPolicy:api.ImageLookupPolicy{Local:false}, DockerImageRepository:"", Tags:map[string]api.TagReference{}}, Status:api.ImageStreamStatus{DockerImageRepository:"127.0.0.1:5000/default/test", Tags:map[string]api.TagEventList{"bar":api.TagEventList{Items:[]api.TagEvent{api.TagEvent{Created:v1.Time{Time:time.Time{sec:63632804584, nsec:433725131, loc:(*time.Location)(0xfe9db60)}}, DockerImageReference:"127.0.0.1:5000/default/test@sha256:56b7f2f33a1d792863c659b4b83f0f7c60cc1530ea1e9a40a85ffcfc5a244991", Image:"sha256:56b7f2f33a1d792863c659b4b83f0f7c60cc1530ea1e9a40a85ffcfc5a244991", Generation:1}, api.TagEvent{Created:v1.Time{Time:time.Time{sec:63632803170, nsec:813666004, loc:(*time.Location)(0xfe9db60)}}, DockerImageReference:"127.0.0.1:5000/default/test@sha256:b9574fbdcd3ae40981970af9f5b8fcda6dddaabc9d7f3fd316d9c54dcf8dad3c", Image:"sha256:b9574fbdcd3ae40981970af9f5b8fcda6dddaabc9d7f3fd316d9c54dcf8dad3c", Generation:1}}, Conditions:[]api.TagEventCondition(nil), PendingCopy:(*api.ImageStreamTagClone)(nil)}}}}
I0611 16:45:42.429633   10669 factory.go:185] Image stream "test" updated.
```

Registry seems to have huge fan out, this is a local test env but I would have expected ACL caching here @mfojtik @legionus

smarterclayton commented:
@bparees see the `oc push binary ...` command.

mfojtik commented:
@smarterclayton is this also implementing this card? https://trello.com/c/dUYS0Hb1/540-5-support-importing-local-docker-images-using-oc-import-image-evg

mfojtik commented:
@soltysh FYI

bparees commented on pkg/cmd/cli/cmd/push/binary.go:
```diff
@@ -0,0 +1,375 @@
+		if err != nil {
+			return fmt.Errorf("--from requires a valid image stream tag or image reference: %v", err)
+		}
+		if len(ref.Registry) > 0 || (len(ref.ID) == 0 && len(ref.Tag) == 0) {
```
looks like parsedockerimagereference will leave the tag empty (vs defaulting it to latest) if unspecified.  are you intending to force people to specify :latest?

bparees commented on pkg/cmd/cli/cmd/push/binary.go:
```diff
@@ -0,0 +1,375 @@
+		if err != nil {
+			return "", nil, 0, fmt.Errorf("the URL passed to filename %q is not valid: %v", s, err)
+		}
+		res, err := http.Get(s)
```
proxy env var support?

smarterclayton commented:
Yes forcing.   http.Get does automatic proxy var support, doesn't it?

On Fri, Jun 16, 2017 at 2:56 PM, Ben Parees <notifications@github.com>
wrote:

> *@bparees* commented on this pull request.
> ------------------------------
>
> In pkg/cmd/cli/cmd/push/binary.go
> <https://github.com/openshift/origin/pull/14470#discussion_r122508520>:
>
> > +	}
> +
> +	to, err := imageapi.ParseDockerImageReference(o.To)
> +	if err != nil {
> +		return fmt.Errorf("--to is not a valid reference: %v", err)
> +	}
> +	if len(to.Registry) > 0 || len(to.ID) > 0 || len(to.Tag) == 0 {
> +		return fmt.Errorf("--to must be NAMESPACE/NAME:TAG or NAME:TAG")
> +	}
> +
> +	if len(o.BaseImage) > 0 {
> +		ref, err := imageapi.ParseDockerImageReference(o.BaseImage)
> +		if err != nil {
> +			return fmt.Errorf("--from requires a valid image stream tag or image reference: %v", err)
> +		}
> +		if len(ref.Registry) > 0 || (len(ref.ID) == 0 && len(ref.Tag) == 0) {
>
> looks like parsedockerimagereference will leave the tag empty (vs
> defaulting it to latest) if unspecified. are you intending to force people
> to specify :latest?
> ------------------------------
>
> In pkg/cmd/cli/cmd/push/binary.go
> <https://github.com/openshift/origin/pull/14470#discussion_r122509635>:
>
> > +	if err != nil {
> +		return err
> +	}
> +	return o.PrintObject([]*resource.Info{info})
> +}
> +
> +func contentsForPathOrURL(s string, in io.Reader, subpaths ...string) (string, io.Reader, int64, error) {
> +	switch {
> +	case s == "-":
> +		return "", in, -1, nil
> +	case strings.Index(s, "http://") == 0 || strings.Index(s, "https://") == 0:
> +		_, err := url.Parse(s)
> +		if err != nil {
> +			return "", nil, 0, fmt.Errorf("the URL passed to filename %q is not valid: %v", s, err)
> +		}
> +		res, err := http.Get(s)
>
> proxy env var support?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/14470#pullrequestreview-44631601>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_pza8kFRX4bN3AYy9r6rZBeiKTQKoks5sEs_wgaJpZM4Nvjte>
> .
>

bparees commented:
sorry I guess it does, last time i looked into this i thought i found otherwise.

```
DefaultTransport is the default implementation of Transport and is used by DefaultClient. It establishes network connections as needed and caches them for reuse by subsequent calls. It uses HTTP proxies as directed by the $HTTP_PROXY and $NO_PROXY (or $http_proxy and $no_proxy) environment variables.
```

smarterclayton commented:
/test all

bparees commented on pkg/cmd/cli/cmd/push/binary.go:
```diff
@@ -0,0 +1,368 @@
+		This command assists users in creating new images that can be used as part of builds
+		or as deployed applications. It accepts a zip or tar.gz file that will become a new 
+		image in an image stream. You may also specify image metadata like the entrypoint, 
+		environment variables, or labels to create a runnable image.
```
there's an implicit assumption here that the archive is going to be extracted when it is used to create the image layer.  might want to make that explicit.  (you get an image containing the contents of the zip, not an image containing the zip)

bparees commented:
still seems like a lot of code for a marginal net gain over:

```
$ oc new-build -D $'FROM scratch\nADD .\n' -n mybuild
$ oc start-build mybuild --from-dir .
```

smarterclayton commented:
Not a marginal net gain.  No pod is executed.  Can be run from any client.  Allows you to update metadata for any existing image and create a new image.

How much java code would you have to write to emulate the flow above for new-build?

bparees commented:
> Not a marginal net gain. No pod is executed. Can be run from any client.

"If we can't make running pods mindlessly simple/cheap/fast we've failed.  If users are trying to avoid running pods to optimize their workflow, we've failed."  -me, i guess.

>  Allows you to update metadata for any existing image and create a new image.

I'll give you this one.  I'd also give you that even the steps i laid out are not going to be trivially obvious to most users, but that's more about the ease of use/UX than the actual functionality.

> How much java code would you have to write to emulate the flow above for new-build?

post a hardcoded buildconfig object w/ the in-line dockerfile and I guess an imagestream for the output.  Nothing about it needs to be customized except maybe the name of the BC and IS. Assuming i've already got the boilerplate api client code in place to make calls, not very much.  For that matter I could use curl to post the two resources. ` oc new-build` is doing very little work in this particular invocation.

the hardest part would be streaming the content, but the client has to do that in both approaches, right?

smarterclayton commented:
On Jul 31, 2017, at 11:57 PM, Ben Parees <notifications@github.com> wrote:

Not a marginal net gain. No pod is executed. Can be run from any client.

"If we can't make running pods mindlessly simple/cheap/fast we've failed.
If users are trying to avoid running pods to optimize their workflow, we've
failed." -me, i guess.

Allows you to update metadata for any existing image and create a new image.

I'll give you this one. I'd also give you that even the steps i laid out
are not going to be trivially obvious to most users, but that's more about
the ease of use/UX than the actual functionality.

How much java code would you have to write to emulate the flow above for
new-build?

post a hardcoded buildconfig object w/ the in-line dockerfile and I guess
an imagestream for the output. Nothing about it needs to be customized
except maybe the name of the BC and IS. Assuming i've already got the
boilerplate api client code in place to make calls, not very much. For that
matter I could use curl to post the two resources. oc new-build is doing
very little work in this particular invocation.


But in the build case I have to map my workflow to a docker file to use a
build pod to create an image.  In this flow, I'm able to layer images
independent of any build process.  The API is also more focused on the
particular goal - I want image X, but with these specific elements
replaced.  Docker files can't clear fields without lots of pain.

This is halfway between "I can run a dockerfile" and "I can upload files to
an http server".  Both have their place, but this also decouples binary
build from its dependence on a particular pod running right now.

the hardest part would be streaming the content, but the client has to do
that in both approaches, right?

And dealing with timeouts and failures.  And the pull delay.  Ultimately
this is also more efficient for anything where you are looking to craft
your own layers.  Binary build is also directly tied to one build
definition.  I can drive multiple off of this flow.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/14470#issuecomment-319260931>,
or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p0aqb9wA0FmiLWGNsrH2g5ryoGwRks5sTqIQgaJpZM4Nvjte>
.

mfojtik commented:
@smarterclayton we need a generated client for this

soltysh commented on pkg/cmd/server/origin/ensure.go:
```diff
@@ -28,7 +30,7 @@ func (c *MasterConfig) ensureOpenShiftSharedResourcesNamespace() {
 		namespace, createErr := c.KubeClientsetInternal().Core().Namespaces().Create(&kapi.Namespace{ObjectMeta: metav1.ObjectMeta{Name: c.Options.PolicyConfig.OpenShiftSharedResourcesNamespace}})
 		if createErr != nil {
-			glog.Errorf("Error creating namespace: %v due to %v\n", c.Options.PolicyConfig.OpenShiftSharedResourcesNamespace, createErr)
+			utilruntime.HandleError(fmt.Errorf("error creating namespace: %v due to %v", c.Options.PolicyConfig.OpenShiftSharedResourcesNamespace, createErr))
```
1. These changes should be a separate commit.
2. This is completely rewritten by David.

soltysh commented on vendor/github.com/docker/distribution/registry/client/blob_writer.go:
```diff
@@ -99,7 +99,7 @@ func (hbu *httpBlobUpload) Write(p []byte) (n int, err error) {
 		return 0, fmt.Errorf("bad range format: %s", rng)
 	}
 
-	return (end - start + 1), nil
```
@dmage @miminar ptal

dmage replied:
lgtm :+1:

https://docs.docker.com/registry/spec/api/#blob-upload

> *On Success: Data Accepted*
>
> ```
> 204 No Content
> Location: /v2/<name>/blobs/uploads/<uuid>
> Range: 0-<offset>
> Content-Length: 0
> Docker-Upload-UUID: <uuid>
> ```
>
> `Range` Range indicating the current progress of the upload.

So, basically, (end - start + 1) is a total size of uploaded data, and this is not what the writer should return.

dmage replied:
:-1: this method should NEVER be called, you have to use ReadFrom.

smarterclayton replied:
Which method are you referring to?  Write has a very specific contract in Go.

dmage replied:
Yes, it has, in general this fix is right and should be upstreamed. But I'm more curious how do you find this? A sequence of writes will produce a sequence of HTTP requests while ReadFrom produces only one HTTP request. And the sequence of HTTP requests not only utilises network less efficient, but may fail on the registry side if it hits S3 (temporary) read-after-update inconsistency. So technically it implements io.Write and it works, but it's better to pretend it's just io.ReaderFrom. Fortunately io.Copy helps with it and uses ReadFrom if it can.

smarterclayton replied:
Oh, it was because we use a MultiReader in order to calculate the digest and tarsum of the layer as we stream it.  MultiReader requires the nested stream to correctly return the write size, otherwise it returns ErrShortWrite

smarterclayton replied:
MultiReader can't use Copy to get ReadFrom.  It's likely few people have ever streamed this content.

dmage replied:
Do you mean MultiWriter? I can't find Multi{Reader,Writer} in this PR, but in general `io.Copy(io.MutliWriter(blobWriter, digester), tarStream)` can be replaced by `blobWriter.ReadFrom(io.TeeReader(tarStream, digester))`.

smarterclayton replied:
clone.go line 467

soltysh commented:
@smarterclayton is this addressing the trello card @mfojtik pointed out or something else. The PR looks reasonable, but I don't like the API. We're currently struggling with the image API badly. I've talked about it with @mfojtik that we would like to clean it up a little bit, especially try to follow upstream patterns. Because currently that's not the case. I'm still interested how much desirable is this functionality, as well.

smarterclayton commented:
> is this addressing the trello card @mfojtik pointed out or something else. 

It's different.  This is about reducing the cost and complexity for a naive client (unprivileged) to do the following three operations:

1. take an image and change the metadata, getting a new image that uses the same layers (change entrypoint, labels, volumes, annotations).
2. create a new scratch image by providing only a blob and single layer
3. create a new image by patching an existing one, changing just a few files

The docker build API can do some of this, but not all of it.  It also requires that you have access to run pods, while this API lets you do this locally without having docker installed.  buildah might allow this in the future, but is too linux specific.  This API would allow the UI to handle this as well.

> The PR looks reasonable, but I don't like the API. We're currently struggling with the image API badly. I've talked about it with @mfojtik that we would like to clean it up a little bit, especially try to follow upstream patterns. Because currently that's not the case. 

Which upstream?  Kube?  I disagree, Kube just has more limited use cases than the image API.  Subresources are pretty open ended.  Docker?  Relevant to some degree, but I'm not as concerned with consistency there.

 > I'm still interested how much desirable is this functionality, as well.

I think this should replace all binary build functionality (or at least, deprecate it) because it allows rebuild.  I think this is going to be the tool that lets end users patch existing images with debug tools, then run it on the cluster.  That's almost impossible to do today.

dmage commented on pkg/image/registry/imagestreamtag/clone.go:
```diff
@@ -0,0 +1,794 @@
+	w := io.MultiWriter(bw, blobhash, pw)
+
+	// upload the bytes
+	n, err := io.Copy(w, r)
```
As we discussed below: `io.Copy(bw, io.TeeReader(r, io.MultiWriter(blobhash, pw))` or `bw.ReadFrom(io.TeeReader(r, io.MultiWriter(blobhash, pw)))`

dmage replied:
And it should be gzipped reader.

Implementation of layer uploading in Moby: https://github.com/moby/moby/blob/d4f6db83c21cfc6af54fffb1f13e8acb7199f96a/distribution/push_v2.go#L433-L455

smarterclayton replied:
Thanks, that's cleaner.

miminar commented on pkg/oc/cli/cmd/push/binary.go:
```diff
@@ -0,0 +1,368 @@
+
+	pushBinaryExample = templates.Examples(`
+		# Upload myblob.zip as a new image in the app:binary image stream tag
+	  $ %[1]s binary --to=app:binary myblob.zip
```
Why not `oc push image`? `binary` could be used for secrets, data for storage volumes or even serialized API objects.

dmage replied:
s/<space><space>/<tab>/

smarterclayton replied:
Hrm - I'd say in openshift the verb "push" belongs most strongly to images.  So in this case push implies image, and binary is the chunk of thing you're pushing.  So like binary build today, we'd have the equivalents to `--from-repo`, `--from-file`, and `--from-dir`.  This is more like `--from-file`.

miminar replied:
Still, I'm quite sure that users will find it hard to tell what `oc binary ...` does just from the name.

smarterclayton replied:
Yeah, the command is marked experimental mostly to get experience with using this for simple tooling as `oc push binary`

miminar commented on pkg/oc/cli/cmd/push/binary.go:
```diff
@@ -0,0 +1,368 @@
+		return fmt.Errorf("--to is not a valid reference: %v", err)
+	}
+	if len(to.Registry) > 0 || len(to.ID) > 0 || len(to.Tag) == 0 {
+		return fmt.Errorf("--to must be NAMESPACE/NAME:TAG or NAME:TAG")
```
Every image push/pull command defaults missing `:TAG` to `:latest`. Why making this special?

smarterclayton replied:
In this case it's an image stream tag, which we never default.  Only docker commands allow defaulting.

miminar commented on pkg/oc/cli/cmd/push/binary.go:
```diff
@@ -0,0 +1,368 @@
+	if err != nil {
+		if errors.IsAlreadyExists(err) {
+			glog.V(2).Infof("Unable to push: %v", err)
+			return fmt.Errorf("a push has already been requested for this tag, use --force to push anyway")
```
I don't see the `Force` being used anywhere.

miminar commented on pkg/oc/cli/cmd/push/binary.go:
```diff
@@ -0,0 +1,368 @@
+			contents, err := os.Open(s)
+			return s, contents, stat.Size(), err
+		}
+		for _, sub := range subpaths {
```
Picking up the first non-directory entry in a given directory seems wrong. The user should be informed that he gave a directory instead of an expected file.

smarterclayton replied:
If subpaths is provided, that's the thing we use to look it up.  This is a generic method being reused here, you'd use that like `contentsForPathOrURL("somedir", nil, "Dockerfile")`.  You get an error if Dockerfile is not used.  If subpaths is empty, you get the not found error.

dmage commented on pkg/image/registry/imagestreamtag/clone.go:
```diff
@@ -0,0 +1,793 @@
+
+	var imageLayer *imageapi.ImageLayer
+	var imageLayerDiffID digest.Digest
+	if r != nil {
```
s/r/layerBody/ ?

smarterclayton replied:
Yeah, should probably be renamed when we're in a context where we are using it as a specific type.

dmage commented on pkg/image/registry/imagestreamtag/clone.go:
```diff
@@ -0,0 +1,793 @@
+		imageConfig.RootFS.DiffIDs = append(imageConfig.RootFS.DiffIDs, diffID.String())
+	}
+
+	// add the scratch layer in if no other layers exist4
```
typo: 4

dmage commented on pkg/image/registry/imagestreamtag/clone.go:
```diff
@@ -0,0 +1,793 @@
+	mediaType, params, err := mime.ParseMediaType(contentType)
+	if err != nil || !strings.HasPrefix(mediaType, "multipart/") {
+		err := errors.NewGenericServerResponse(http.StatusNotAcceptable, "post", h.r.rest.gr, h.name, "", 0, false)
+		err.ErrStatus.Message = fmt.Sprintf("The only accepted media type is '%s' with boundary specified. The first part must have content type application/json and contain an ImageStreamTagClone object. The second part must be a Docker file layer with content type '%s'.", formDataContentType, manifest.DockerV2Schema2LayerMediaType)
```
s/'%s'/%q/ ?

dmage commented on pkg/image/registry/imagestreamtag/clone.go:
```diff
@@ -0,0 +1,793 @@
+	imageConfig *imageapi.DockerImageConfig,
+	base *imageapi.Image,
+	layer *imageapi.ImageLayer,
+	diffID digest.Digest,
```
diffID is not unused

dmage commented:
> this should replace all binary build functionality
Does it mean that all pushes from builders will be proxies through the api server?

smarterclayton commented:
Other way around.  A binary build today uses attach to push a blob to the
pod.  Next iteration would push an image to the registry.  Both (in this
flow) go through the apiserver.

In the future, it should be easier for a client to create the image on
their side with the external docker registry API, but that's an
optimization (still has to go through a router or other ingress).


On Aug 7, 2017, at 8:03 PM, Oleg Bulatov <notifications@github.com> wrote:

this should replace all binary build functionality
Does it mean that all pushes from builders will be proxies through the api
server?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/14470#issuecomment-320812939>,
or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p_rrkSPZcSl-R1Zxg61VWw8hyoezks5sV6XmgaJpZM4Nvjte>
.

smarterclayton commented:
After talking with @soltysh renamed this to "instantiate" operation to be consistent with build configs and deployment configs, where instantiate takes options that are different from the base object (you don't PUT deployment configs to /deploymentconfigs/instantiate) and then instantiates the thing you are posting or pushing to, which in this case is a new image tagged into the destination.

smarterclayton commented on pkg/image/apis/image/helper.go:
```diff
@@ -494,9 +494,9 @@ func ImageWithMetadata(image *Image) error {
-		for i, j := 0, len(image.DockerImageLayers)-1; i < j; i, j = i+1, j-1 {
-			image.DockerImageLayers[i], image.DockerImageLayers[j] = image.DockerImageLayers[j], image.DockerImageLayers[i]
-		}
+		// for i, j := 0, len(image.DockerImageLayers)-1; i < j; i, j = i+1, j-1 {
```
@legionus when i look at a schema2 image in docker hub (mysql:latest) - oldest layer is first in the manifest layers, top layer is last.  Up until now we didn't need this (because the manifest is upstream pullthrough), I realized as part of this that the ordering was inverted in our version of the manifest.  Can you confirm that schema1 and schema2 *should* do have different orders, or help me identify why this is sufficient to generate a new image correctly?

legionus replied:
@smarterclayton https://github.com/docker/distribution/blob/master/docs/spec/manifest-v2-2.md#image-manifest-field-descriptions

* layers array
  The layer list is ordered starting from the base image (opposite order of schema1).

legionus replied:
@smarterclayton I think this schema2 reordering is wrong. schema1 and schema2 should have same order in our database.

smarterclayton replied:
Ok, will spawn a separate issue with tests for this.  This only impacts when we need to generate a new manifest / content from an existing image, so I don't know if there's an impact to our users that requires a backport.

legionus commented on pkg/image/importer/credentials.go:
```diff
@@ -183,6 +184,15 @@ func basicCredentialsFromKeyring(keyring credentialprovider.DockerKeyring, targe
 		}
+
+		// try removing the canonical ports
+		if strings.HasSuffix(target.Host, ":443") || strings.HasSuffix(target.Host, ":80") {
```
I think that here it is necessary to check the `target.Schemа`. Isn't it ?

smarterclayton replied:
Yeah, I think there are several cases:

schema = https
schema = http
schema = none
port = 443
port = 80
port = unspecified

For unspecified or canonical schema or port, the credentials should be sent to a secured server.      I think when schema is unspecified and the canonical port is used, we should infer the schema.  We should probably be careful not to send credentials to HTTP endpoints unless the user requests it, but on the other hand a user might reasonably expect a host like `foo.example.com` to be sent to the server.

legionus commented on pkg/image/registry/imagestreamtag/instantiate.go:
```diff
@@ -0,0 +1,809 @@
+	blobs := repo.Blobs(ctx)
+	bw, err := blobs.Create(ctx)
+	if err != nil {
+		return distribution.Descriptor{}, "", nil, fmt.Errorf("unable to begin uploading blob: %v", err)
```
Please add `bw.Cancel(ctx)` here.

dmage replied:
https://github.com/docker/distribution/blob/5f6282db7d65e6d72ad7c2cc66310724a57be716/blobs.go#L235

> Cancel implementations should allow multiple calls even after a commit that result in a no-op. This allows use of Cancel in a defer statement

legionus replied:
@dmage yes. but defer is not used here :)

legionus commented on pkg/image/registry/imagestreamtag/instantiate.go:
```diff
@@ -0,0 +1,809 @@
+	}
+
+	// TODO: use configurable digests
+	layerdigest := digest.NewDigestFromBytes(digest.SHA256, layerhash.Sum(make([]byte, 0, layerhash.Size())))
```
`digest.Canonical` ?
https://github.com/openshift/origin/blob/master/vendor/github.com/docker/distribution/digest/digester.go#L24

legionus commented on pkg/image/registry/imagestreamtag/instantiate.go:
```diff
@@ -0,0 +1,809 @@
+// This comes from github.com/docker/distribution/manifest/schema1/config_builder.go; there is
+// a non-zero embedded timestamp; we could zero that, but that would just waste storage space
+// in registries, so let’s use the same values.
+var gzippedEmptyLayer = []byte{
```
We already have it in few places. Maybe better to make it as a separate module ?

https://github.com/openshift/origin/search?utf8=%E2%9C%93&q=gzippedEmptyTar

smarterclayton replied:
I considered reusing the value in container/image but was unsure how deeply we want to depend on that library in this code path.

legionus commented on pkg/oc/cli/cmd/push/binary.go:
```diff
@@ -0,0 +1,363 @@
+
+func (o *PushBinaryOptions) Validate() error {
+	if len(o.To) == 0 {
+		return fmt.Errorf("--to=IMAGESTREAMTAG is a required argument")
```
@dmage @smarterclayton It would be fun to add support for manifest lists here :)
I know that now we do not support them, but still.

smarterclayton replied:
Yeah, we're going to have to deal with this sooner or later.  I may end up removing this command from this PR (so just the API) so we can discuss how we want to expose it separately.  The API is fairly self contained, while the commands get a lot into "what we want to do with this feature".  

If manifest lists can be updated later, then we have more flexibility to alter how copy works when used against a manifest list vs a single item.

smarterclayton commented on pkg/oc/cli/cmd/push/binary.go:
```diff
@@ -0,0 +1,363 @@
+
+func (o *PushBinaryOptions) Validate() error {
+	if len(o.To) == 0 {
+		return fmt.Errorf("--to=IMAGESTREAMTAG is a required argument")
```
Yeah - for this API we could start by generating them automatically if we had to under the covers.  Do all images in a manifest list have to be pushed at the same time (i.e. manifest list -> array of manifests)?

legionus replied:
@smarterclayton As far as I know no. But the big question is which image you decide to change because it depends on `platform` object [1]. Especially it is not clear how to handle `features` as it is an array of some strings (unspecified). So for one architecture there can be more than one version of the image [2].

[1] https://github.com/docker/distribution/blob/master/docs/spec/manifest-v2-2.md#manifest-list-field-descriptions

[2] https://docs.docker.com/datacenter/dtr/2.3/guides/release-notes/
> Support was added for handling manifest-lists for multi-architecture images. This lets you manage images for different operating systems (eg. Linux and Windows) and CPU architectures (eg. x86_64 and s390x) under a single tag.

legionus replied:
Also the logic of image selection in docker client was not documented anywhere.

dmage commented on pkg/oc/cli/cmd/push/binary.go:
```diff
@@ -0,0 +1,363 @@
+		glog.V(4).Infof("Detected input stream as zip")
+		return pr
+	default:
+		tr := tar.NewReader(br)
```
The call `tr.Next()` will read some data from `br` and that data will not appear in the result stream. It should be `tar.NewReader(bytes.NewBuffer(buf))`.

dmage commented on pkg/oc/cli/cmd/push/binary.go:
```diff
@@ -0,0 +1,363 @@
+			gw := gzip.NewWriter(pw)
+			go func() {
+				if _, err := io.Copy(gw, br); err != nil {
+					pw.CloseWithError(err)
```
This can be called unconditionally. And the next call `pw.Close()` can be removed.

dmage replied:
Nevermind, in case of success you need to close gw to have a properly finished gzip archive.

dmage commented on pkg/oc/cli/cmd/push/binary.go:
```diff
@@ -0,0 +1,363 @@
+	switch {
+	case s == "-":
+		return "", in, -1, nil
+	case strings.Index(s, "http://") == 0 || strings.Index(s, "https://") == 0:
```
`strings.HasPrefix` will better express the intent.

dmage commented on pkg/oc/cli/cmd/push/binary.go:
```diff
@@ -0,0 +1,363 @@
+		if err != nil {
+			return "", nil, 0, err
+		}
+		return "", res.Body, -1, nil
```
As it is returned as io.Reader, it will be never closed.

smarterclayton commented:
I'm going to split the commands out from this PR.  The changes to image lookup are in #15945.  This will just be the API bits and tests for those.

smarterclayton commented:
Updated with all comments addressed and a set of extended tests added.  The CLI command will not be delivered as part of this, but will be part of follow up around binary build.  I added `os` to all docker images so we start having the metadata available (but images would have to be migrated to capture the field since it is at create time)

smarterclayton commented:
Still debating what I want to do with this.  I considered moving it to its
own image command `oc image layer` and having it depend on layer reuse
working correctly, which should always be true for a given cluster.

On Sun, Feb 25, 2018 at 6:29 PM, OpenShift Bot <notifications@github.com>
wrote:

> Issues go stale after 90d of inactivity.
>
> Mark the issue as fresh by commenting /remove-lifecycle stale.
> Stale issues rot after an additional 30d of inactivity and eventually
> close.
> Exclude this issue from closing by commenting /lifecycle frozen.
>
> If this issue is safe to close now please do so with /close.
>
> /lifecycle stale
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/14470#issuecomment-368355446>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p7Eadkl2vL4cb8E4rQIR-c8aS48Fks5tYezmgaJpZM4Nvjte>
> .
>
